File: LeetCode/StudyRoadMap/ArraysAndHashing/validAnagram.js
/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
const isAnagram = function (s, t) {
    // hash map with key as char and value as count
    // for s and t

    if (s.length !== t.length) return false;

    let [sMap, tMap] = [new Map(), new Map()];

    for (let i = 0; i < s.length; i++) {
        // count the occurance of each char in s and t
        sMap.set(s[i], sMap.get(s[i]) + 1 || 1);
        tMap.set(t[i], tMap.get(t[i]) + 1 || 1);
    }

    // compare the maps
    for (let [key, value] of sMap) {
        if (value !== tMap.get(key)) return false;
    }
    return true;


};

console.log(isAnagram("anagram", "nagaram"));
console.log(isAnagram("car", "rat"));

// s = "anagram", t = "nagaram" => true
// aaangrm

File: LeetCode/StudyRoadMap/ArraysAndHashing/valid-anagram.py
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False
        # create two hash containing the count of letters in str
        countS, countT = {}, {}
        # since both str are same length
        for i in range(len(s)):
            # if the key doesnt exist in the hashmap default value 0
            countS[s[i]] = 1 + countS.get(s[i], 0)
            countT[t[i]] = 1 + countT.get(t[i], 0)

        print(countS)
        print(countT)
        # iterate the hash maps
        for c in countS:
            if countS[c] != countT.get(c, 0):
                return False
        return True

    def isAnagramSort(self, s: str, t: str) -> bool:
        return sorted(s) == sorted(t)


test = Solution()
print(test.isAnagram('anagram', 'nagaram'))
print(test.isAnagramSort('anagram', 'nagaram'))
